packages/gatsby/src/bootstrap/requires-writer.js

```diff
@@ -80,20 +102,15 @@ const getMatchPaths = pages => {
 
   return matchPathPages
     .sort((a, b) => {
-      // Paths with wildcards should appear after those without.
-      const wildcardOrder = b.wildcard - a.wildcard
-      if (wildcardOrder !== 0) {
-        return wildcardOrder
-      }
-
       // The higher the score, the higher the specificity of our matchPath
       const order = b.score - a.score
       if (order !== 0) {
         return order
       }
 
-      // if specificity is the same we use the array index
-      return b.index - a.index
+      // if specificity is the same we do lexigraphic comparison of path to ensure
+      // deterministic order regardless of order pages where created
+      return a.matchPath.localeCompare(b.matchPath)
```

Comment:
This also need some eyes on (this is not exactly something needed for this fix, it just should mean that `matchPaths.json` order is always the same regardless of order of pages)